Mathlib/Logic/Equiv/Fin.lean

```diff
@@ -546,3 +547,32 @@ instance subsingleton_fin_zero : Subsingleton (Fin 0) :=
 instance subsingleton_fin_one : Subsingleton (Fin 1) :=
   finOneEquiv.subsingleton
 #align subsingleton_fin_one subsingleton_fin_one
+
+def Fin.equivOfSuccEquiv (h : Fin (n+1) ≃ Fin (m+1)) : Fin n ≃ Fin m :=
+  Equiv.removeNone <| calc
+    _ ≃ Fin (n+1) := (finSuccEquiv _).symm
+    _ ≃ Fin (m+1) := h
+    _ ≃ _         := finSuccEquiv _
+
+/-!
+  ## Injectivity of Fin
+  As a corollary of these equivalences, we get that `Fin` is injective
+-/
+
+theorem Fin.inj_equiv (eqv : Fin n ≃ Fin m) : n = m := by
+  induction n generalizing m <;> cases m
+  case zero.zero      => rfl
+  case zero.succ      => exact Fin.elim0 <| eqv.symm ⟨0, by simp⟩
+  case succ.zero      => exact Fin.elim0 <| eqv ⟨0, by simp⟩
+  case succ.succ ih _ => rw [ih]; apply Fin.equivOfSuccEquiv eqv
+
+theorem Fin.inj : Fin n = Fin m → n = m :=
```

Comment:
We already have this result, it's [`fin_injective`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Fintype/Card.html#fin_injective)

Reply:
Whoops, so we do. I expected it to be called some variation of `Fin.inj` or `Fin.injective`, so when those did not give any hits on the doc search I assumed we did not have the result yet.
Do you reckon we could/should make [`Fin.cast_eq_cast'`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Fintype/Card.html#Fin.cast_eq_cast') a simp-lemma? The need for injectivity of `Fin` usually comes up when dealing with `cast`, so making that a simp-lemma would make it a lot more discoverable

Reply:
I think needing injectivity of `Fin` is an XY problem; injectivity of fin can certainly help you get out of a mess, but I think you can always avoid getting into that mess in the first place.